Adds an oss-fuzz harness to fuzz AssemblyStack API calls for parsing …

bshastry opened the pull request:
…StrictAssembly and optimizing to EVM (Constantinople)

(Closes #5900 )

### Description

This PR

- creates a new oss-fuzz test harness for fuzzing the parsing and optimization of StrictAssembly code
- adds a dictionary for strict assembly and an oss-fuzz config file

bshastry committed:
Adds oss-fuzz harnesses to fuzz AssemblyStack API calls for parsing/optimizing StrictAssembly and generating EVM bytecode.

chriseth commented on test/tools/ossfuzz/config/strict_assembly.dict:
```diff
@@ -0,0 +1,118 @@
+"("
+")"
+", "
+":"
```
Probably better to leave that out - it is only required for types.

bshastry replied:
not sure I follow. The following Solidity assembly snippet from [here](https://solidity.readthedocs.io/en/v0.5.3/assembly.html) contains all of cited tokens

```
        assembly {
            // retrieve the size of the code, this needs assembly
            let size := extcodesize(_addr)
            // allocate output byte array - this could also be done without assembly
            // by using o_code = new bytes(size)
            o_code := mload(0x40)
            // new "memory end" including padding
            mstore(0x40, add(o_code, and(add(add(size, 0x20), 0x1f), not(0x1f))))
            // store length in memory
            mstore(o_code, size)
            // actually retrieve the code, this needs assembly
            extcodecopy(_addr, add(o_code, 0x20), 0, size)
```

chriseth replied:
sure, but `:` only appears as part of `:=`.

chriseth commented on test/tools/ossfuzz/config/strict_assembly.dict:
```diff
@@ -0,0 +1,118 @@
+")"
+", "
+":"
+" dup1 "
```
explicit dupN opcodes are not allowed in strict assembly.

bshastry replied:
Considering `jump(r)` resulted in a failed assertion, shouldn't we leave these (including `dup`) for now with the hope that we fix other potential errors due to wrong expectations encoded in the compiler?

chriseth replied:
We mostly want to find problems with the optimizer and because of that the grammar should focus on generating input that passes the parser, shouldn't it?

bshastry replied:
I agree. But this `{jump(r)}` causes an internal compiler exception inside the parser. It should be a graceful parser error (return false instead of throwing an assertion), shouldn't it?

chriseth commented on test/tools/ossfuzz/config/strict_assembly.dict:
```diff
@@ -0,0 +1,118 @@
+" dup9 "
+" pc "
+" stop "
+" swap1 "
```
swap is also not allowed.

bshastry replied:
same comment as for `dup`

chriseth commented on test/tools/ossfuzz/config/strict_assembly.dict:
```diff
@@ -0,0 +1,118 @@
+"calldatacopy("
+"calldataload("
+" calldatasize "
+" caller "
```
In strict assembly, all opcodes must be used as functions, even if they have no arguments, so I would propose to use `caller()` here, `pc()` above and similar for other opcodes.

bshastry replied:
okay.

chriseth commented on test/tools/ossfuzz/config/strict_assembly.dict:
```diff
@@ -0,0 +1,118 @@
+"sub("
+" timestamp "
+"xor("
+"assembly {"
```
This is not part of strict assembly.

bshastry replied:
removed

chriseth commented on test/tools/ossfuzz/config/strict_assembly.dict:
```diff
@@ -0,0 +1,118 @@
+"jump("
+"jumpi("
+"keccak256("
+"log0("
```
`let` is missing

bshastry replied:
added

chriseth commented on test/tools/ossfuzz/config/strict_assembly.dict:
```diff
@@ -0,0 +1,118 @@
+"smod("
+"sstore("
+"staticcall("
+"sub("
```
`switch`, `case`, `default` and literal numbers (some examples) and strings are missing.

bshastry replied:
done.

chriseth commented on test/tools/ossfuzz/strictasm_opt_ossfuzz.cpp:
```diff
@@ -0,0 +1,36 @@
+
+extern "C" int LLVMFuzzerTestOneInput(uint8_t const* _data, size_t _size)
+{
+	if (_size > 450)
```
Hm, that's not a lot. Any reason for that particular number?

bshastry replied:
This is a heuristic. Let me explain.

When no size limit is imposed, the fuzzer generates very large inputs (because we seed the fuzzer with test cases over a kilobyte or more). Some of these large inputs contain so many syntactical issues that the parser throws a fatalerror because the number of parsing errors encountered is too many.

I noticed that `75%` of our assembly test cases are under `450` bytes and felt that the marginal utility of the other `25%` is not worth making the parser trip because of badly formatted input.

chriseth replied:
ok, sounds good!

axic commented on test/tools/ossfuzz/strictasm_opt_ossfuzz.cpp:
```diff
@@ -0,0 +1,36 @@
+		return 0;
+
+	stack.optimize();
+	return 0;
```
Do we want to test for codegen too? `.assemble()`

bshastry replied:
I had this question too. @chriseth feels that fuzzing `.assemble()` will most likely trigger a (known) assertion for exceeding the number of stack slots that is currently supported (16 I think?)

axic replied:
Ideally I'd run two versions of the fuzzer, one with optimization and one without. Without optimizations it definitely should run compile, with the optimizer on we can disable that for now.

bshastry replied:
@axic When you say "run compile", are you referring to the `compileEVM(adapter, false, _optimize);` API?

If so, I notice that this is called within `AssemblyStack::assemble(Machine _machine, bool _optimize)`. So, would the following harness be something that you have in mind for the other version (no optimization, just parse and assemble)?

```
	if (!stack.parseAndAnalyze("source", input))
		return 0;

	stack.assemble();
```

axic replied:
When I say compile I mean calling `stack.assemble()`.

chriseth replied:
We can compile, but then we should handle stack problems as not an error.

bshastry replied:
@chriseth Would handling stack problems as not an error (i.e., ignoring them for the purpose of fuzz testing other code portions of the assembler) be as simple as removing the following assertion inside `EVMObjectCompiler::run`?

```
yulAssert(transform.stackErrors().empty(), "Stack errors present but not thrown.");
```

chriseth replied:
No, this assertion should not be reachable. Please try it out with code like this one here: https://github.com/ethereum/solidity/blob/2c75c08a892574b3668589e4b76f5e914bbbeb96/test/libyul/CompilabilityChecker.cpp#L95

bshastry replied:
@chriseth tested with the linked inputs, no assertions were raised. Also, fuzzing the `assembly()` API on the corpus collected from `optimize()` fuzzing turns up no assertions as well.

bshastry replied:
> Ideally I'd run two versions of the fuzzer, one with optimization and one without. Without optimizations it definitely should run compile

Updated the PR with two versions: one parses and optimizes, the other parses and assembles w/o optimization

bshastry replied:
> No, this assertion should not be reachable. Please try it out with code like this one here:

Code like this triggers the `StackTooDeepError` exception
```
{function f(x_1,x_2,x_3,x_4,x_5,x_6,x_7,x_8,x_9,x_10,x_11,x_12,x_13,x_14,x_15,x_16,x_17) {} }
```

I am curious why the test that you linked from `solidity/test/libyul/CompilabilityChecker.cpp` does not do this. My reasoning is that, in the latter case, the excessive stack slots are from inside the function scope (not parameter list) and the optimization removes them thereby saving stack slots.

On the other hand, the optimizer cannot (should not?) optimize away function parameters.

chriseth replied:
My guess is that if you run the full suite, it just removes the function, if it is not called anywhere.

bshastry replied:
Update: I now catch `StackTooDeepError` to filter bug reports like above.

chriseth commented on test/tools/ossfuzz/strictasm_assembly_ossfuzz.cpp:
```diff
@@ -0,0 +1,46 @@
+using namespace yul;
+using namespace std;
+
+extern "C" int LLVMFuzzerTestOneInput(const uint8_t* _data, size_t _size)
```
```suggestion
extern "C" int LLVMFuzzerTestOneInput(uint8_t const* _data, size_t _size)
```

bshastry replied:
`_data` is not a const pointer but a pointer to a const array of bytes.

chriseth replied:
Yes indeed. Otherwise it would be `uint8_t * const _data`. The order of type and const-specifier is irrelevant for C/C++ and our coding style is "const-specifier after type".

bshastry replied:
This always confuses me :) will fix.

chriseth commented:
Completely forgot about this one. What's the state here, @bshastry ?

bshastry commented:
Got sidetracked :)

 Will update the PR after a fresh rebase.

bshastry commented:
I rebased this PR. It is now ready to be merged. I also removed irrelevant tokens such as `dupN`, `pushN`, and `swapN` from the strict assembly dictionary.
